### ISSUE TITLE
Question about X-Call-Id headers

### ISSUE THREAD

**Issue by Space-Banana-42** [openfaas/faas#1624]
## My actions before raising this issue
- [x] Followed the [troubleshooting guide](https://docs.openfaas.com/deployment/troubleshooting/)
- [x] Read/searched [the docs](https://docs.openfaas.com/)
- [x] Searched [past issues](/issues)

## Expected Behaviour

"X-Call-Id" header added to response body of function call though OpenFaaS Gateway.

## Current Behaviour

We updated the OpenFaaS gateway from 0.17.0 to 0.18.18 and found "X-Call-Id" header is missing in response body in 0.18.18. We also tested the with 0.20.7, which also doesn't give us "X-Call-Id".

## Possible Solution

Change the order of these two lines:
https://github.com/openfaas/faas/blob/11ff356ccea32b89b9da556c18305cd7c5ea9ba5/gateway/main.go#L144
https://github.com/openfaas/faas/blob/5c2bc2e0c7c95d06e6b202fd2924d05365f6269c/gateway/main.go#L184
In 0.17.0, ```faasHandlers.QueuedProxy = handlers.MakeCallIDMiddleware(faasHandlers.QueuedProxy)``` actually happens before ```functionProxy := faasHandlers.Proxy```.

If "X-Call-Id " don't need to apply to response header, I think it might be better to just remove ```faasHandlers.QueuedProxy = handlers.MakeCallIDMiddleware(faasHandlers.QueuedProxy)``` since it doesn't take any effect.

## Steps to Reproduce (for bugs)

1. Deploy latest OpenFaaS with helm2 on kubernetes
2. Deploy any sync function
3. send a request to that function
4. "X-Call-Id" is missing from response header. 

## Context

We update the OpenFaaS gateway from 0.17.0 to 0.18.18 and found "X-Call-Id" header is missing in response header for gateway 0.18.18.
At very beginning, we don't expect our customer to use "X-Call-Id" in response header, but since it is already there in gateway 0.17.0, and our customer already using that header, when we try to upgrade OpenFaaS to new version, our customer complain about this issue :-(

## Your Environment

* FaaS-CLI version ( Full output from: `faas-cli version` ):

* Docker version `docker version` (e.g. Docker 17.0.05 ):

* Are you using Kubernetes or faasd?
Kubernetes
* Operating System and version (e.g. Linux, Windows, MacOS):

* Code example or link to GitHub repo or gist to reproduce problem:

* Other diagnostic information / logs from [troubleshooting guide](https://docs.openfaas.com/deployment/troubleshooting)

## Next steps

You may [join Slack](https://docs.openfaas.com/community) for community support.

**Comment by Space-Banana-42** [comment]
Made a small test:
```
package main

import (
	"fmt"
	"github.com/gorilla/mux"
	"net/http"
	"time"
)

func MakeCallIDMiddleware(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		if len(r.Header.Get("X-Call-Id")) == 0 {
			callID := "this is a test"
			r.Header.Add("X-Call-Id", callID)
			w.Header().Add("X-Call-Id", callID)
		}

		r.Header.Add("X-Start-Time", fmt.Sprintf("%d", start.UTC().UnixNano()))
		w.Header().Add("X-Start-Time", fmt.Sprintf("%d", start.UTC().UnixNano()))

		next(w, r)
	}
}

func TestHandler(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(http.StatusOK)
	fmt.Fprintf(w, "biu: %v\n", "biu")
}

func main() {
	fmt.Println("Hello, World!")
	faasHandlersProxy := TestHandler

	functionProxy := faasHandlersProxy
	faasHandlersProxy = MakeCallIDMiddleware(faasHandlersProxy)
	r := mux.NewRouter()
	r.HandleFunc("/test", functionProxy)
	http.Handle("/", r)
	s := &http.Server{
		Addr:           "127.0.0.1:8000",
		ReadTimeout:    15 * time.Second,
		WriteTimeout:   15 * time.Second,
		MaxHeaderBytes: http.DefaultMaxHeaderBytes, // 1MB - can be overridden by setting Server.MaxHeaderBytes.
		Handler:        r,
	}

	s.ListenAndServe()
}
```
If change the order of line 35 and 36, it does change the response header.

**Comment by alexellis** [comment]
We would accept a PR for this including the unit test.

Thank you for letting us know.

**Comment by Space-Banana-42** [comment]
@alexellis happy to create one

**Comment by alexellis** [comment]
I just tested the gateway with version 0.20.7 and cannot reproduce the error. I used faasd.

Initial request

```
curl http://192.168.0.15:8080/async-function/nodeinfo -d "" -H "X-Callback-Url: http://192.168.0.83:8080"
```

And the result with call-id:

```
alex@am1 faas % nc -l 192.168.0.83 8080
POST / HTTP/1.1
Host: 192.168.0.83:8080
User-Agent: Go-http-client/1.1
Content-Length: 88
Content-Type: text/plain; charset=utf-8
Date: Sat, 06 Feb 2021 09:15:53 GMT
Etag: W/"58-C7bZymf8DX6ndd8MCHrwBYXhajU"
X-Call-Id: 81786f49-b0e5-4ab7-9826-208bc77f4b02
X-Duration-Seconds: 0.203057
X-Function-Name: nodeinfo
X-Function-Status: 200
Accept-Encoding: gzip
Connection: close

Hostname: localhost

Arch: arm
CPUs: 4
Total mem: 926MB
Platform: linux
Uptime: 6960658
```

**Comment by alexellis** [comment]
Are you expecting a call-id from synchronous calls? I don't think that is something we have intentionally supported at any time apart from with async calls.

See more on call IDs:
https://docs.openfaas.com/reference/async/#callback-request-headers

**Comment by Space-Banana-42** [comment]
> Are you expecting a call-id from synchronous calls? I don't think that is something we have intentionally supported at any time apart from with async calls.
> 
> See more on call IDs:
> https://docs.openfaas.com/reference/async/#callback-request-headers

Thanks for your reply! Yes, we are expectiong "X-Call-Id" from synchronous call. Sorry if I misunderstood, but I do notice in README.md of [gateway](https://github.com/openfaas/faas/blob/master/gateway/README.md) that "An "X-Call-Id" header is applied to every incoming call through the gateway and is usable for tracing and monitoring calls. We use a UUID for this string."

If that is just a feature for async call, might it be possible that it applies for every incoming call? Thanks!

**Comment by alexellis** [comment]
It turns out that this was regression. I've fixed it in #1627
